cmd/promtool/analyze.go

```diff
@@ -0,0 +1,543 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+var (
+	errNotNativeHistogram = fmt.Errorf("not a native histogram")
+	errNotEnoughData      = fmt.Errorf("not enough data")
+)
+
+type AnalyzeHistogramsConfig struct {
+	metricType string
+	duration   time.Duration
+	end        string
+	step       time.Duration
+	matchers   []string
+}
+
+// run retrieves metrics that look like conventional histograms (i.e. have _bucket
+// suffixes) or native histograms, depending on metricType flag.
+func (c *AnalyzeHistogramsConfig) run(url *url.URL, roundtripper http.RoundTripper) error {
+	if c.metricType != "classichistograms" && c.metricType != "nativehistograms" {
+		return fmt.Errorf("histogram type is %s, must be 'classichistograms' or 'nativehistograms'", c.metricType)
+	}
+
+	ctx := context.Background()
+
+	api, err := newAPI(url, roundtripper, nil)
+	if err != nil {
+		return err
+	}
+
+	var endTime time.Time
+	if c.end != "" {
+		endTime, err = parseTime(c.end)
+		if err != nil {
+			return fmt.Errorf("error parsing end time '%s': %w", c.end, err)
+		}
+	} else {
+		endTime = time.Now()
+	}
+	startTime := endTime.Add(-c.duration)
+
+	if c.metricType == "nativehistograms" {
+		// histoMetrics, err := queryMetadataForHistograms(ctx, api, "", "1000")
+		histoMetrics, err := queryMetricNames(ctx, api, c.matchers, startTime, endTime)
+		if err != nil {
+			return err
+		}
+		return c.getStatsFromMetrics(ctx, api, startTime, endTime, os.Stdout, identity, identity, calcNativeBucketStatistics, histoMetrics)
```

Comment:
Let me first describe back to you how I understand this so that you can correct me if I'm wrong:

1. `queryMetricNames` gets all the names for each of the provided matchers.
2. Next, diving into `getStatsFromMetrics`, you iterate through all the metric names and get the full vector for each via `count_over_time` (in a separate query for each name).
3. Then you iterate through that vector and query samples of each element in a range query, with an estimated step based on the `count_over_time` result (or the user provided step, if it exists). (This again multiplies the number of required queries.)
4. Finally, you run the bucket analysis on each set of samples coming back from the range query.

For one, I think this might do the wrong thing under special circumstances. Imagine the user provides `--match='my_histogram{instance="foo"}`. Step (1) yields `my_histogram`, so that step (2) might yield way more `my_histogram` elements than just `my_histogram{instance="foo"}`, for example it would also yield `my_histogram{instance="bar"}`.

But I also have another thought, expanding on my idea from the first review iteration: Could we perhaps do all of the above in one step? And without asking the user for a step (or scrape interval or whatever we call it)?

This is the idea:

For each of the provided matchers, run a simple instant query `<matcher>[<duration>]` with `<end>` as the evaluation time (defaulting to "now"). For example, if the matcher is `my_histogram{instance="foo"}'` and the user-provided duration is 1h, the query would look like `my_histogram{instance="foo"}[1h]`. This would then return a matrix matching precisely the series the user provided matcher matches, with the raw samples from the requested duration, with their raw original timestamp and spacing, so no need to do any artificial request for a scrape interval at all.

This would be much simpler in general, and also requires just one query per matcher instead of the many many queries in the flow above

Reply:
I tried to do this but couldn't figure out how to make it work with only 1 query per matcher, and still work with wildcard matching, e.g. this matcher wouldn't work: `{__name__=~\"cortex_request_.*\"}[1m]`

One problem is that the metric `my_histogram{instance="foo"}` might have many series which may not be time-aligned, which is why the current flow processes each series one-by-one.

You're right that the example of `--match='my_histogram{instance="foo"}` wouldn't return the desired result, but it seems rather complicated to allow matching specific series instead of just metrics. Is this a use-case we need to support?

Reply:
> I tried to do this but couldn't figure out how to make it work with only 1 query per matcher, and still work with wildcard matching, e.g. this matcher wouldn't work: `{__name__=~\"cortex_request_.*\"}[1m]`

Hmm, it should work. Could you tell me what error messages you get?

> One problem is that the metric `my_histogram{instance="foo"}` might have many series which may not be time-aligned, which is why the current flow processes each series one-by-one, and also the buckets and usage patterns may be entirely different.

Yes, but that's the way it is, right? The range query artificially creates regularly spaced samples, but they are still coming from the raw samples, now maybe with duplications or omissions. My idea would be to to one big instant query (with a range vector inside) and then go through the matrix that comes back, series by series, meaning it is still one-by-one in effect, but really on the raw samples.

> You're right that the example of `--match='my_histogram{instance="foo"}` wouldn't return the desired result, but it seems rather complicated to allow matching specific series instead of just metrics. Is this a use-case we need to support?

On a large Prometheus, just querying a single metric name (AKA metric family) might return a _lot_ of series. (At SoundCloud, we had a Prometheus server that was 80% full of just one histogram metric family.) Dumping most of the TSDB through a query into promtool might overwhelm both sides, the server and promtool…

There might still be the use case of "I actually want to dump the whole TSDB and analyse every single histogram in it", which will require to do the work in batches, but that's much more complicated to implement than my proposal here, so I would suggest to do the MVP first and then iterate on it later for other use cases.

Reply:
Thanks for the suggestions, have reduced the number of queries. Still have to do extra queries for classic histograms to first get the matching series, and then the set of buckets separately, but native histograms have minimal queries.